### ISSUE TITLE
Make UDP ports identify themselves when stopping

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5308]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#5308|
|Type|enhancement|
|Created|2011-10-11 11:54:53Z|
|Branch|https://github.com/twisted/twisted/tree/udp-port-ident-5308|

When a TCP port is stopped, it logs a message:

```
(TCP Port 39770 Closed)
```

When a UDP port is stopped, it logs the less informative:

```
(Port 46434 Closed)
```

One can infer that it is a UDP port since it did not say that it is a TCP port, but it would be better just to say.

One possible trick to watch out for, though, is that udp sets its log str to include the string "UDP".  A log message like:

```
[DatagramProtocol (UDP)] (UDP Port 46434 Closed)
```

might thus result with this change.  This is hardly the end of the world, though.

<details><summary>Searchable metadata</summary>

```
trac-id__5308 5308
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_udp_port_ident_5308 branches/udp-port-ident-5308
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__easy easy
time__1318334093000000 1318334093000000
changetime__1318794115000000 1318794115000000
version__None None
owner__exarkun exarkun
cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [32869]) Branching to 'udp-port-ident-5308'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [32870]) Add tests for the UDP port start and stop log messages; make the stop message include "UDP"; also fix the UDP stopListening implementation to not use the global reactor

refs [#5308](https://github.com/twisted/twisted/issues/5308)
refs [#5299](https://github.com/twisted/twisted/issues/5299)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [32871]) Replicate those fixes exactly in iocpreactor

refs [#5308](https://github.com/twisted/twisted/issues/5308)
refs [#5299](https://github.com/twisted/twisted/issues/5299)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [32872]) Incorporate zseil's test from [[#5299](https://github.com/twisted/twisted/issues/5299)](https://github.com/twisted/twisted/issues/5299)

refs [#5308](https://github.com/twisted/twisted/issues/5308)
refs #5299

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

I did a few things, fortunately the branch is still small:

  1. Added tests for UDP start and stop log messages.  I changed the start message to be consistent with the TCP start message - that is, to use the `ILoggingContext` implementation on the datagram protocol, if there is one.
  1. Changed the stop message to include "UDP" in both the iocp and posixbase implementations.
  1. To actually make the stop test pass, I folded in zseil's patch from [#5299](https://github.com/twisted/twisted/issues/5299).  This makes udp port shutdown use the right reactor for scheduling, necessary for reactor builder style tests to work.  His change also included the deletion of some unnecessary code in `udp.Port.connectionLost` (in both posixbase and iocpreactor).
  1. Cleaned up some incidental code in `test_tcp.py` which I noticed while looking at that code to write the udp tests, and refactored the log observer helper from test_tcp so it could be re-used in test_udp.

This branch should resolve both this ticket and [#5299](https://github.com/twisted/twisted/issues/5299).  [Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/udp-port-ident-5308) seem to look okay.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_7"></a>|@jesstess set owner to @exarkun|
|-|-|

Thanks for working on this, exarkun. One tiny comment:

 * "L{DatagramProtocol.stopProtocol} is called asynchronously (ie, not re-entrantly) in when C{stopListening}" ==> "L{DatagramProtocol.stopProtocol} is called asynchronously (ie, not re-entrantly) when C{stopListening}"

Other than that, looks great, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set status to `closed`|
|-|-|

(In [32883]) Merge udp-port-ident-5308

Author: zseil, exarkun
Reviewer: itamar, jesstess
Fixes: [#5299](https://github.com/twisted/twisted/issues/5299)
Fixes: [#5308](https://github.com/twisted/twisted/issues/5308)

Change UDP ports to:

  * Use the `DatagramProtocol`'s `ILoggingContext` implementation for startup messages
  * log their type (ie, UDP) when they are closed
  * Use the reactor passed to their initializer to schedule connectionLost, rather than the global reactor
